File: Overlord.Core/Overlord.Core/BombardmentSystem.cs
using Overlord.Core.Models;

namespace Overlord.Core;

/// <summary>
/// Platform-agnostic bombardment system.
/// Handles orbital bombardment, structure destruction, and civilian casualties.
/// </summary>
public class BombardmentSystem
{
    private readonly GameState _gameState;
    private readonly Random _random;

    /// <summary>
    /// Event fired when bombardment starts.
    /// Parameters: (planetID, strength)
    /// </summary>
    public event Action<int, int>? OnBombardmentStarted;

    /// <summary>
    /// Event fired when structure is destroyed.
    /// Parameters: (structureID, planetID)
    /// </summary>
    public event Action<int, int>? OnStructureDestroyed;

    /// <summary>
    /// Event fired when civilian casualties occur.
    /// Parameters: (planetID, casualties)
    /// </summary>
    public event Action<int, int>? OnCivilianCasualties;

    public BombardmentSystem(GameState gameState, Random? random = null)
    {
        _gameState = gameState ?? throw new ArgumentNullException(nameof(gameState));
        _random = random ?? new Random();
    }

    /// <summary>
    /// Executes orbital bombardment of a planet.
    /// </summary>
    /// <param name="planetID">Target planet ID</param>
    /// <param name="attackerFaction">Attacking faction</param>
    /// <returns>Bombardment result if successful, null if failed</returns>
    public BombardmentResult? BombardPlanet(int planetID, FactionType attackerFaction)
    {
        // Get planet
        if (!_gameState.PlanetLookup.TryGetValue(planetID, out var planet))
            return null;

        // Cannot bombard own planets
        if (planet.Owner == attackerFaction)
            return null;

        // Must control orbit (no enemy craft)
        if (!HasOrbitalControl(planetID, attackerFaction))
            return null;

        // Calculate bombardment strength (Battle Cruisers × 50)
        var battleCruisers = GetBattleCruisersInOrbit(planetID, attackerFaction);
        if (battleCruisers.Count == 0)
            return null; // No bombardment craft

        int bombardmentStrength = battleCruisers.Count * 50;

        OnBombardmentStarted?.Invoke(planetID, bombardmentStrength);

        // Destroy structures
        var destroyedStructures = DestroyStructures(planet, bombardmentStrength);

        // Cause civilian casualties
        int casualties = CauseCasualties(planet, bombardmentStrength);

        // Reduce morale by 20%
        int oldMorale = planet.Morale;
        planet.Morale = Math.Max(0, planet.Morale - 20);

        // Create result
        var result = new BombardmentResult
        {
            PlanetID = planetID,
            BombardmentStrength = bombardmentStrength,
            StructuresDestroyed = destroyedStructures.Count,
            DestroyedStructureIDs = destroyedStructures,
            CivilianCasualties = casualties,
            NewMorale = planet.Morale
        };

        return result;
    }

    /// <summary>
    /// Destroys structures based on bombardment strength.
    /// </summary>
    /// <param name="planet">Target planet</param>
    /// <param name="bombardmentStrength">Bombardment strength</param>
    /// <returns>List of destroyed structure IDs</returns>
    private List<int> DestroyStructures(PlanetEntity planet, int bombardmentStrength)
    {
        var destroyedStructureIDs = new List<int>();

        // Calculate number of structures to destroy (max 3)
        int structuresToDestroy = Math.Min(bombardmentStrength / 100, 3);

        // Get vulnerable structures (exclude Docking Bays and Orbital Defense)
        var vulnerableStructures = planet.Structures
            .Where(s => s.Type != BuildingType.DockingBay &&
                       s.Type != BuildingType.OrbitalDefense &&
                       s.Status == BuildingStatus.Active)
            .ToList();

        if (vulnerableStructures.Count == 0)
            return destroyedStructureIDs;

        // Randomly select structures to destroy
        for (int i = 0; i < structuresToDestroy && vulnerableStructures.Count > 0; i++)
        {
            int randomIndex = _random.Next(vulnerableStructures.Count);
            var structure = vulnerableStructures[randomIndex];

            // Destroy structure
            planet.Structures.Remove(structure);
            destroyedStructureIDs.Add(structure.ID);
            vulnerableStructures.RemoveAt(randomIndex);

            OnStructureDestroyed?.Invoke(structure.ID, planet.ID);
        }

        _gameState.RebuildLookups();
        return destroyedStructureIDs;
    }

    /// <summary>
    /// Causes civilian casualties (10% of bombardment strength).
    /// </summary>
    /// <param name="planet">Target planet</param>
    /// <param name="bombardmentStrength">Bombardment strength</param>
    /// <returns>Number of casualties</returns>
    private int CauseCasualties(PlanetEntity planet, int bombardmentStrength)
    {
        int casualties = (int)Math.Floor(bombardmentStrength * 0.1f); // 10% of strength
        casualties = Math.Min(casualties, planet.Population); // Cannot kill more than population

        planet.Population -= casualties;

        OnCivilianCasualties?.Invoke(planet.ID, casualties);
        return casualties;
    }

    /// <summary>
    /// Checks if faction controls orbit (no enemy craft).
    /// </summary>
    /// <param name="planetID">Planet ID</param>
    /// <param name="faction">Faction checking control</param>
    /// <returns>True if faction controls orbit</returns>
    public bool HasOrbitalControl(int planetID, FactionType faction)
    {
        var craftInOrbit = _gameState.Craft
            .Where(c => c.PlanetID == planetID && !c.IsDeployed)
            .ToList();

        // No enemy craft = orbital control
        var enemyCraft = craftInOrbit.Where(c => c.Owner != faction).ToList();
        return enemyCraft.Count == 0;
    }

    /// <summary>
    /// Gets Battle Cruisers in orbit for a faction.
    /// </summary>
    /// <param name="planetID">Planet ID</param>
    /// <param name="faction">Faction</param>
    /// <returns>List of Battle Cruisers</returns>
    private List<CraftEntity> GetBattleCruisersInOrbit(int planetID, FactionType faction)
    {
        return _gameState.Craft
            .Where(c => c.PlanetID == planetID &&
                       c.Owner == faction &&
                       c.Type == CraftType.BattleCruiser &&
                       !c.IsDeployed)
            .ToList();
    }
}

File: Overlord.Phaser/src/core/BombardmentSystem.ts
import { GameState } from './GameState';
import { FactionType, BuildingType, BuildingStatus, CraftType } from './models/Enums';
import { BombardmentResult } from './models/CombatModels';
import { PlanetEntity } from './models/PlanetEntity';
import { CraftEntity } from './models/CraftEntity';
// Note: Structure type is inferred from planet.structures array

/**
 * Platform-agnostic bombardment system.
 * Handles orbital bombardment, structure destruction, and civilian casualties.
 */
export class BombardmentSystem {
  private readonly gameState: GameState;
  private readonly random: () => number;

  /**
   * Event fired when bombardment starts.
   * Parameters: (planetID, strength)
   */
  public onBombardmentStarted?: (planetID: number, strength: number) => void;

  /**
   * Event fired when structure is destroyed.
   * Parameters: (structureID, planetID)
   */
  public onStructureDestroyed?: (structureID: number, planetID: number) => void;

  /**
   * Event fired when civilian casualties occur.
   * Parameters: (planetID, casualties)
   */
  public onCivilianCasualties?: (planetID: number, casualties: number) => void;

  constructor(gameState: GameState, randomFunc?: () => number) {
    if (!gameState) {
      throw new Error('gameState cannot be null or undefined');
    }

    this.gameState = gameState;
    this.random = randomFunc || Math.random;
  }

  /**
   * Executes orbital bombardment of a planet.
   * @param planetID Target planet ID
   * @param attackerFaction Attacking faction
   * @returns Bombardment result if successful, null if failed
   */
  public bombardPlanet(planetID: number, attackerFaction: FactionType): BombardmentResult | null {
    // Get planet
    const planet = this.gameState.planetLookup.get(planetID);
    if (!planet) {
      return null;
    }

    // Cannot bombard own planets
    if (planet.owner === attackerFaction) {
      return null;
    }

    // Must control orbit (no enemy craft)
    if (!this.hasOrbitalControl(planetID, attackerFaction)) {
      return null;
    }

    // Calculate bombardment strength (Battle Cruisers × 50)
    const battleCruisers = this.getBattleCruisersInOrbit(planetID, attackerFaction);
    if (battleCruisers.length === 0) {
      return null; // No bombardment craft
    }

    const bombardmentStrength = battleCruisers.length * 50;

    this.onBombardmentStarted?.(planetID, bombardmentStrength);

    // Destroy structures
    const destroyedStructures = this.destroyStructures(planet, bombardmentStrength);

    // Cause civilian casualties
    const casualties = this.causeCasualties(planet, bombardmentStrength);

    // Reduce morale by 20%
    planet.morale = Math.max(0, planet.morale - 20);

    // Create result
    const result = new BombardmentResult();
    result.planetID = planetID;
    result.bombardmentStrength = bombardmentStrength;
    result.structuresDestroyed = destroyedStructures.length;
    result.destroyedStructureIDs = destroyedStructures;
    result.civilianCasualties = casualties;
    result.newMorale = planet.morale;

    return result;
  }

  /**
   * Destroys structures based on bombardment strength.
   * @param planet Target planet
   * @param bombardmentStrength Bombardment strength
   * @returns List of destroyed structure IDs
   */
  private destroyStructures(planet: PlanetEntity, bombardmentStrength: number): number[] {
    const destroyedStructureIDs: number[] = [];

    // Calculate number of structures to destroy (max 3)
    const structuresToDestroy = Math.min(Math.floor(bombardmentStrength / 100), 3);

    // Get vulnerable structures (exclude Docking Bays and Orbital Defense)
    const vulnerableStructures = planet.structures.filter(
      s =>
        s.type !== BuildingType.DockingBay &&
        s.type !== BuildingType.OrbitalDefense &&
        s.status === BuildingStatus.Active,
    );

    if (vulnerableStructures.length === 0) {
      return destroyedStructureIDs;
    }

    // Randomly select structures to destroy
    for (let i = 0; i < structuresToDestroy && vulnerableStructures.length > 0; i++) {
      const randomIndex = Math.floor(this.random() * vulnerableStructures.length);
      const structure = vulnerableStructures[randomIndex];

      // Destroy structure
      const structureIndex = planet.structures.indexOf(structure);
      if (structureIndex !== -1) {
        planet.structures.splice(structureIndex, 1);
      }
      destroyedStructureIDs.push(structure.id);
      vulnerableStructures.splice(randomIndex, 1);

      this.onStructureDestroyed?.(structure.id, planet.id);
    }

    this.gameState.rebuildLookups();
    return destroyedStructureIDs;
  }

  /**
   * Causes civilian casualties (10% of bombardment strength).
   * @param planet Target planet
   * @param bombardmentStrength Bombardment strength
   * @returns Number of casualties
   */
  private causeCasualties(planet: PlanetEntity, bombardmentStrength: number): number {
    let casualties = Math.floor(bombardmentStrength * 0.1); // 10% of strength
    casualties = Math.min(casualties, planet.population); // Cannot kill more than population

    planet.population -= casualties;

    this.onCivilianCasualties?.(planet.id, casualties);
    return casualties;
  }

  /**
   * Checks if faction controls orbit (no enemy craft).
   * @param planetID Planet ID
   * @param faction Faction checking control
   * @returns True if faction controls orbit
   */
  public hasOrbitalControl(planetID: number, faction: FactionType): boolean {
    const craftInOrbit = this.gameState.craft.filter(c => c.planetID === planetID && !c.isDeployed);

    // No enemy craft = orbital control
    const enemyCraft = craftInOrbit.filter(c => c.owner !== faction);
    return enemyCraft.length === 0;
  }

  /**
   * Gets Battle Cruisers in orbit for a faction.
   * @param planetID Planet ID
   * @param faction Faction
   * @returns List of Battle Cruisers
   */
  private getBattleCruisersInOrbit(planetID: number, faction: FactionType): CraftEntity[] {
    return this.gameState.craft.filter(
      c =>
        c.planetID === planetID &&
        c.owner === faction &&
        c.type === CraftType.BattleCruiser &&
        !c.isDeployed,
    );
  }
}
